components/entities/api/src/main/java/org/phenotips/entities/PrimaryEntity.java

```diff
@@ -53,7 +55,14 @@
      *
      * @return a valid document reference
      */
-    DocumentReference getDocument();
+    DocumentReference getDocumentReference();
+
+    /**
+     * Returns the document where the entity is stored.
+     *
+     * @return a valid document reference
+     */
+    XWikiDocument getDocument();
```

Comment:
I don't think changing the return type of `DocumentReference getDocument()` (to `XWikiDocument getDocument()`) is a good choice, because it breaks code that already uses that method. The `DocumentReference getDocument()` API has already been in use for a few years (e.g. [here](https://github.com/phenotips/phenotips/blob/61b9c338cb3ee3b2bb064d5b68f9b1affa2946fc/components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/ClinicalStatusController.java#L74)), so this change is likely to affect a lot of downstream code. I already encountered code breakage resulting from this change in a downstream Gene42 extension. 

The fact that this PR mostly consists of small changes within PT to make things work with the `getDocument()` API change is a good indication that lots of downstream code will likely need to be tweaked as well. I think the harm of breaking downstream code in this case far outweighs the benefit of introducing a slightly more convenient version of the method.

Fortunately, we can easily maintain backwards compatibility while still adding this more convenient method. How about keeping `DocumentReference getDocument()` and adding a new method `XWikiDocument getXWikiDocument()` instead?